gum/arch-mips/gummipsrelocator.c

```diff
@@ -0,0 +1,530 @@
+}
+
+void
+gum_mips_relocator_reset (GumMipsRelocator * relocator,
+                           gconstpointer input_code,
+                           GumMipsWriter * output)
+{
+  guint i;
+
+  relocator->input_start = input_code;
+  relocator->input_cur = input_code;
+  relocator->input_pc = GUM_ADDRESS (input_code);
+  for (i = 0; i != GUM_MAX_INPUT_INSN_COUNT; i++)
+  {
+    cs_insn * insn = relocator->input_insns[i];
+    if (insn != NULL)
+    {
+      cs_free (insn, 1);
+      relocator->input_insns[i] = NULL;
+    }
+  }
+  relocator->output = output;
+
+  relocator->inpos = 0;
+  relocator->outpos = 0;
+
+  relocator->delay_slot_pending = FALSE;
+  relocator->eob = FALSE;
+  relocator->eoi = FALSE;
+}
+
+void
+gum_mips_relocator_free (GumMipsRelocator * relocator)
+{
+  gum_mips_relocator_reset (relocator, relocator->input_start,
+      relocator->output);
+
+  g_free (relocator->input_insns);
+
+  cs_close (&relocator->capstone);
+}
+
+static guint
+gum_mips_relocator_inpos (GumMipsRelocator * self)
+{
+  return self->inpos % GUM_MAX_INPUT_INSN_COUNT;
+}
+
+static guint
+gum_mips_relocator_outpos (GumMipsRelocator * self)
+{
+  return self->outpos % GUM_MAX_INPUT_INSN_COUNT;
+}
+
+static void
+gum_mips_relocator_increment_inpos (GumMipsRelocator * self)
+{
+  self->inpos++;
+  g_assert_cmpint (self->inpos, >, self->outpos);
+}
+
+static void
+gum_mips_relocator_increment_outpos (GumMipsRelocator * self)
+{
+  self->outpos++;
+  g_assert_cmpint (self->outpos, <=, self->inpos);
+}
+
+guint
+gum_mips_relocator_read_one (GumMipsRelocator * self,
+                              const cs_insn ** instruction)
+{
+  cs_insn ** insn_ptr, * insn;
+
+  if (self->eoi && !self->delay_slot_pending)
+    return 0;
+
+  insn_ptr = &self->input_insns[gum_mips_relocator_inpos (self)];
+
+  if (*insn_ptr != NULL)
+  {
+    cs_free (*insn_ptr, 1);
+    *insn_ptr = NULL;
+  }
+
+  if (cs_disasm (self->capstone, self->input_cur, 4, self->input_pc, 1,
+      insn_ptr) != 1)
+  {
+    return 0;
+  }
+
+  insn = *insn_ptr;
+
+  switch (insn->id)
+  {
+    case MIPS_INS_J:
+      self->eob = TRUE;
+      self->delay_slot_pending = TRUE;
+      self->eoi = TRUE;
+      break;
+    case MIPS_INS_JR:
+      self->eob = TRUE;
+      self->delay_slot_pending = TRUE;
+      self->eoi = TRUE;
+      break;
+    case MIPS_INS_BGEZAL:
+    case MIPS_INS_BGEZALL:
+    case MIPS_INS_BLTZAL:
+    case MIPS_INS_BLTZALL:
+    case MIPS_INS_JAL:
+    case MIPS_INS_JALR:
+      self->eob = TRUE;
+      self->delay_slot_pending = TRUE;
+      self->eoi = FALSE;
+      break;
+    case MIPS_INS_BEQ:
+    case MIPS_INS_BEQL:
+    case MIPS_INS_BGEZ:
+    case MIPS_INS_BGEZL:
+    case MIPS_INS_BGTZ:
+    case MIPS_INS_BGTZL:
+    case MIPS_INS_BLEZ:
+    case MIPS_INS_BLEZL:
+    case MIPS_INS_BLTZ:
+    case MIPS_INS_BLTZL:
+    case MIPS_INS_BNE:
+    case MIPS_INS_BNEL:
+      self->eob = TRUE;
+      self->delay_slot_pending = TRUE;
+      self->eoi = FALSE;
```

Comment:
Would be good to set `delay_slot_pending` last here and above, for consistency with the proposed header change.